src/main/scala/cromwell/binding/expression/WdlStandardLibraryFunctions.scala

```diff
@@ -9,25 +9,78 @@ import scala.util.{Failure, Success, Try}
 trait WdlStandardLibraryFunctions extends WdlFunctions[WdlValue] {
   private def fail(name: String) = Failure(new UnsupportedOperationException(s"$name() not implemented yet"))
 
+  private def extractObjectOrArray(params: Seq[Try[WdlValue]]) = for {
+    singleArgument <- extractSingleArgument(params)
+    contents <- Success(fileContentsToString(singleArgument))
+    wdlObjects <- WdlObject.fromTsv(contents)
+  } yield wdlObjects
+
   protected def stdout(params: Seq[Try[WdlValue]]): Try[WdlFile] = fail("stdout")
   protected def stderr(params: Seq[Try[WdlValue]]): Try[WdlFile] = fail("stderr")
-  protected def read_lines(params: Seq[Try[WdlValue]]): Try[WdlArray] = fail("read_lines")
-  protected def read_tsv(params: Seq[Try[WdlValue]]): Try[WdlArray] = fail("read_tsv")
-  protected def read_map(params: Seq[Try[WdlValue]]): Try[WdlMap] = fail("read_map")
-  protected def read_object(params: Seq[Try[WdlValue]]): Try[WdlObject] = fail("read_objects")
-  protected def read_objects(params: Seq[Try[WdlValue]]): Try[WdlArray] = fail("read_objects")
+
+  /**
+    * Read all lines from the file referenced by the first parameter and return an Array[String]
+    */
+  protected def read_lines(params: Seq[Try[WdlValue]]): Try[WdlArray] = {
+    for {
+      singleArgument <- extractSingleArgument(params)
+      lines = fileContentsToString(singleArgument).split("\n").map{WdlString}
+    } yield WdlArray(WdlArrayType(WdlStringType), lines)
+  }
+
+ protected def read_map(params: Seq[Try[WdlValue]]): Try[WdlMap] = {
+    for {
+      singleArgument <- extractSingleArgument(params)
+      contents <- Success(fileContentsToString(singleArgument))
+      wdlMap <- WdlMap.fromTsv(contents)
+    } yield wdlMap
+  }
+
+  protected def read_object(params: Seq[Try[WdlValue]]): Try[WdlObject] = {
+    extractObjectOrArray(params) map {
+      case array if array.length == 1 => array.head
+      case _ => throw new IllegalArgumentException("read_object yields an Object and thus can only read 2-rows TSV files. Try using read_objects instead.")
+    }
+  }
+
+  def read_objects(params: Seq[Try[WdlValue]]): Try[WdlArray] =
+    extractObjectOrArray(params) map { WdlArray(WdlArrayType(WdlObjectType), _) }
+
+  /**
+    * Try to read a string from the file referenced by the specified `WdlValue`.
+    */
+  protected def read_string(params: Seq[Try[WdlValue]]): Try[WdlString] = {
+    for {
+      singleArgument <- extractSingleArgument(params)
+      string = fileContentsToString(singleArgument)
+    } yield WdlString(string.trim)
+  }
+
+  /**
+    * Read a file in TSV format into an Array[Array[String]]
+    */
+  protected def read_tsv(params: Seq[Try[WdlValue]]): Try[WdlArray] = {
+    for {
+      singleArgument <- extractSingleArgument(params)
+      contents <- Success(fileContentsToString(singleArgument))
+      wdlArray <- Success(WdlArray.fromTsv(contents))
```

Comment:
I agree for contents which definitely does throw exceptions.
For wdlArray, the fromTsv already returns a Try so maybe just don't use = instead of <- ?